repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/keystonev2/authorizationstandalone/AuthorizationFilterMultiTenantTest.groovy

```diff
@@ -0,0 +1,369 @@
+
+        and: "It should have made it to the origin service"
+        mc.handlings.size() == 1
+
+        and: "Tenant ID header should be removed."
+        def xTenantIdHdrs = mc.handlings[0].request.headers.findAll("x-tenant-id")
+        xTenantIdHdrs.size() == 0
+
+        and: "Roles should not be reduced."
+        def rolesRequestHeaders = mc.handlings[0].request.getHeaders().findAll("X-Roles").join(",").split(",")
+        rolesRequestHeaders as Set == ["role1", "role2", "role3", "role4", "racker"] as Set
+        def mapRolesRequestHeaders = mc.handlings[0].request.getHeaders().findAll("X-Map-Roles")
+        mapRolesRequestHeaders.size() == 1
+        base64Decode(mapRolesRequestHeaders[0]) == mapRoles
+    }
+
+    def "User with Multiple Tenants and URI extraction should receive a 200 OK response"() {
+        given: "a configured tenant"
+        def tenantId = Math.abs(random.nextInt())
+
+        when: "User sends a request through repose"
+        MessageChain mc = deproxy.makeRequest(
+            url: """$reposeEndpoint/extract/$tenantId/ss""",
+            method: "GET",
+            headers: [
+                "X-Catalog": xCatalog,
+                "X-Map-Roles": base64Encode("""{"tenant1":["role1","role2","role3"],"$tenantId":["role3"],"repose/domain/roles":["role4"]}"""),
+                "X-Tenant-Id": "tenant1;q=0.5",
+                "X-Roles": "role1,role2,role3,role4"])
+
+        then: "User should receive a 200 response"
+        mc.receivedResponse.code as Integer == SC_OK
+
+        and: "It should have made it to the origin service"
+        mc.handlings.size() == 1
+
+        and: "Tenant should be extracted from URI and replace existing."
+        def xTenantIdHdrs = mc.handlings[0].request.headers.findAll("x-tenant-id")
+        xTenantIdHdrs.size() == 1
+        xTenantIdHdrs[0] == """$tenantId;q=0.8"""
+
+        and: "Roles should be reduced to only the tenant's and the global"
+        def rolesRequestHeaders = mc.handlings[0].request.getHeaders().findAll("X-Roles").join(",").split(",")
+        rolesRequestHeaders as Set == ["role3", "role4"] as Set
+        def mapRolesRequestHeaders = mc.handlings[0].request.getHeaders().findAll("X-Map-Roles")
+        mapRolesRequestHeaders.size() == 1
+        def mapRolesDecode = base64Decode(mapRolesRequestHeaders[0]).split("[\\[\\]{}:,\"]") as Set
+        if (mapRolesDecode.contains("")) {
+            mapRolesDecode.remove("")
+        }
+        mapRolesDecode == [tenantId as String, "role3", "repose/domain/roles", "role4"] as Set
+    }
+
+    def "User with Multiple Tenants and invalid URI extraction should receive a 401 UNAUTHORIZED response"() {
+        given: "a configured tenant"
+        def tenantId = Math.abs(random.nextInt())
+
+        when: "User sends a request through repose"
+        MessageChain mc = deproxy.makeRequest(
+            url: """$reposeEndpoint/extract/${Math.abs(random.nextInt())}/ss""",
```

Comment:
There's a small chance that this test will randomly fail. If `nextInt` gives the same value twice in a row this test fails. Maybe just add one to the first value you got to guarantee there's never a collision?